trpcultivate_phenotypes/tests/src/Kernel/TripalImporter/TraitImporterProcessValidationTest.php

```diff
@@ -1477,4 +1477,202 @@ public function testProcessDuplicateTraitsFailures(array $failures, array $expec
+        'expected_message' => "Expected to find the key \'failedItems\' in the validation result array.",
+      ],
+    ];
+
+    // #1: 'valid' key has been set to TRUE.
+    $scenarios[] = [
+      [
+        'case' => 'Case 1',
+        'valid' => TRUE,
+        'failedItems' => [
+          'item' => 'failed',
+        ],
+      ],
+      [
+        'expected_exception' => TRUE,
+        'expected_message' => "Expected the validation result to contain a value of FALSE for the key 'valid' since it should only reach this point if validation failed.",
+      ],
+    ];
+
+    // #2: 'failedItems' key contains a value of an empty array.
+    $scenarios[] = [
+      [
+        'case' => 'Case 2',
+        'valid' => FALSE,
+        'failedItems' => [],
+      ],
+      [
+        'expected_exception' => TRUE,
+        'expected_message' => "Expected the validation result to have content for the key 'failedItems', but it was set to an empty array.",
+      ],
+    ];
+
+    // #3: 'failedItems' key contains a value of string, not an array.
+    $scenarios[] = [
+      [
+        'case' => 'Case 3',
+        'valid' => FALSE,
+        'failedItems' => 'item that failed',
+      ],
+      [
+        'expected_exception' => TRUE,
+        'expected_message' => "Expected the validation result to contain an array for the key 'failedItems', but it did not.",
+      ],
+    ];
+
+    // #4: Validation result array not faulty, no exceptions triggered.
+    $scenarios[] = [
+      [
+        'case' => 'Case 4',
+        'valid' => FALSE,
+        'failedItems' => [
+          'item' => 'failed',
+        ],
+      ],
+      [
+        'expected_exception' => FALSE,
+        'expected_message' => 'NONE',
+      ],
+    ];
+
+    return $scenarios;
+  }
+
+  /**
+   * Tests the method that checks the integrity of the validation status array.
+   *
+   * @param array $validation_result
+   *   The validation status array returned by a single validator and where
+   *   applicable, a single row of the input file (ie. a row-level validator).
+   *   It is expected to contain the following keys:
+   *   - 'case': a developer-focused string describing the case checked.
+   *   - 'valid': FALSE to indicate that validation failed.
+   *   - 'failedItems': An array of items that failed.
+   * @param array $expectations
+   *   An array of expectations that we want to find in the resulting rendered
+   *   output. Each array has the following keys:
+   *   - 'expected_exception': TRUE or FALSE if an exception is expected to
+   *     occur for this scenario.
+   *   - 'expected_message': The message expected by the exception
+   *     being triggered.
+   *
+   * @dataProvider provideFaultyValidationStatusArray
+   */
+  public function testCheckValidationStatusArray(array $validation_result, array $expectations) {
+
+    $exception_caught = FALSE;
+    $exception_message = 'NONE';
+    try {
+      $this->importer->checkValidationStatusArray($validation_result);
```

Comment:
I also agree with [Reynolds comment above](https://github.com/TripalCultivate/TripalCultivate-Phenotypes/pull/139#issuecomment-2622499007). In order to use this approach in your current test, I would simply pass in a made-up validator name here and then below when you check the exception message you add another assertStringContainsString() that confirms this made-up validator name is in the exception message.

This approach also gives you maximum coverage without having to trigger an exception for checkValidationStatusArray for each validator.

Reply:
My reason for my initial approach is to also provide access to the row numbers once I start working on the row-level process methods. If I change it, I'll have to make row number an optional parameter in checkValidationStatusArray, and I seem to have an aversion to making methods with optional parameters 🙈 I hope there are not additional details we might want to pass in...

But this IS really tempting to keep the process methods less cluttered AND avoid testing an extra exception in each process method. I will try it out with a row-level one next and let you both know how it goes 😃

Reply:
Instead of making a number of different optional values you could make a single one: array $context that could have both the validator name, row number and any other context for messages we want in the future 🤔 Alternatively, you could make a single required parameter which is string $message_prefix. This way row level validators could include both the validator name and line number but other validators could provide different information.